src/grid/mod.rs

```diff
@@ -479,7 +474,7 @@ impl<'a, T> BidirectionalIterator for GridIterator<'a, T> {
         let num_cols = self.grid.num_cols();
 
         match self.cur {
-            Point { line, col: Column(0) } if line == self.top => None,
+            Point { line, col: Column(0) } if line == self.grid.len() - 1 => None,
```

Comment:
I've changed the `prev()` and `next()` iterators to iterate over the complete grid here, it seems like this works (tests don't fail) and for everything else the DisplayIterator should be used. However there's probably a good reason why this has been done this way, can you maybe elaborate on this @jwilm?

Reply:
Looks like a bug :man_shrugging:

Reply:
`self.top` was used to avoid extra conversions of the form `Line -> usize`. Is it possible that we can do a bounds check when selection is active without needing to change this code?

Reply:
It's possible, but the behavior of this code seems faulty to me. The iterator over the raw grid shouldn't stop as soon as it hits the viewport boundary, should it? Based on my understanding the viewport should be irrelevant for selection.

Also semantic selection is the only consumer of `prev()`, just `next()` is re-used in multiple places.

Reply:
Oh yeah, that sounds right. I'm starting to confuse myself with the different iterators we have which aren't clear about what they are supposed to provide. In some cases, we have something like a `GridViewportIterator` which isn't named as such.

Anyway, sounds like you have the right idea here.

Reply:
FWIW, termite allows semantic selections to span above the viewport, for example when a URL is broken onto two lines.

Reply:
I'll probably run a round of benchmarks just to make sure there wasn't some performance reason for doing things the way they are done right now, that should help provide an additional bit of safety.